### ISSUE TITLE
Arquillian cube should wait for all the containers to be up and running

### ISSUE THREAD

**Issue by vinsguru** [arquillian/arquillian-cube#680]
Unable to use Arquillian Graphene page objects with Arquillian Cube. Even before containers are ready, Arquillian Graphene/drone tries inject the page objects.

---------------

My cube compose file is as shown below.

```
selenium-hub:
  image: selenium/hub
  container_name: selenium-hub
  ports:
    - "4444:4444"
chrome:
  image: selenium/node-chrome
  links:
    - selenium-hub:hub
firefox:
  image: selenium/node-firefox
  links:
    - selenium-hub:hub
```
Arquillian.xml config for docker and drone.

```
 <extension qualifier="docker">
  <property name="serverVersion">1.27</property>
  <property name="serverUri">tcp://10.11.12.13:2375</property>
  <property name="dockerContainersFile">docker-compose.yml</property>
</extension>
 
 <extension qualifier="webdriver">
      <property name="remoteAddress">http://10.11.12.13:4444/wd/hub</property>
      <property name="browser">chrome</property> 
      <property name="remote">true</property>
      <property name="remoteReusable">true</property>
 </extension>
```
TestNG test using selenium grid.
```
	@Page
        Google google;

       @Test
	public void canGoogleSuggest(){
		google.getSearchWidget().searchFor("Arquillian");
		google.getSuggestions().show();
		Assert.assertEquals(google.getSuggestions().getCount(), 4);	
	}
```
I get the exception as grid does to seem to have been yet ready!!

> org.openqa.selenium.WebDriverException: Error forwarding the new session Empty pool of VM for setup Capabilities

If I add some wait, it works fine.

	@Test
	public void canGoogleSuggest(){
                Thread.sleep(1000);
		Google google = Graphene.goTo(Google.class);
		google.getSearchWidget().searchFor("Arquillian");
		google.getSuggestions().show();
		Assert.assertEquals(google.getSuggestions().getCount(), 4);	
	}

**Comment by lordofthejars** [comment]
Hi currently Cube starts containers following next strategy: first all containers that have no links, then containers that depends on containers already started and so on so forth. So in your docker compose `firefox` and `chrome` and then `selenium-hub`. And how Cube knows that  a container is up and running? By checking that the binded port is opened in OS level and is able to receive requests.

In case of browser this can be a bit tricky to detect. I recommend you read this link http://arquillian.org/arquillian-cube/#_cube_docker_drone_graphene_integration where we explain how to use Arquillian drone/graphene with Docker (Arquillian Cube).

Of course feel free to continue the conversation in this issue if that chapter does not fix your problem then we can work together to find a solution. Also if you want to see how this is implemented in Cube you can check this module https://github.com/arquillian/arquillian-cube/tree/master/docker/drone 


Obviously thank you very much for using Arquillian Cube.

**Comment by vinsguru** [comment]
@lordofthejars ,

This is the order in which I see the containers are starting.

1. selenium-hub
2.chrome
3.firefox

The above order is perfectly fine. chrome and firefox containers once created, should register themselves to the selenium-hub. So that hub can route the request to chrome/firefox.

The issue seems to be happening within the gap between the chrome/firefox container creation and registration to the hub. so it is nothing to do with Drone/Graphene. (this registration might take a second) 

Is there any way for us to include the await strategy in the docker-compose file?

**Comment by vinsguru** [comment]
@lordofthejars ,

You guys seem to have considered all the possible combinations of issues one might face! 
The below config solved the issue!!

  <property name="cubeSpecificProperties">
    chrome:
      await:
        strategy: sleeping
        sleepTime: 2 s   
</property>

You guys ROCK!!
Thanks.

**Comment by lordofthejars** [comment]
Thanks for using it :)

**Comment by lordofthejars** [comment]
You might be interested on this https://github.com/arquillian/arquillian-cube/pull/682 to be released in 1.2.0 which will avoid you having to use `cubeSpecificProperties`
